### DISCUSSION TITLE
do not throw error if env variable (ie. mnemonic) is missing

### DISCUSSION THREAD

**Discussion by SvenMeyer** [PaulRBerg/hardhat-template#66] [Ideas]
I just had the case where I send the link to my repo out for audit, and all I got was a complaint (concerning tests) that they did not work.

When running the tests with hardhat, no mnemonic is needed, so should not cause any error if env is not set.

Adding more and more chains, and verifying pages, it becomes more and more likely that you/somebody has not / does not want to set a certain env.

Ideally the presence would only be checked when the variable would really be needed. I.e. I want to deploy on BSC, then I would only need a mnemonic for that chain.

Any idea for a smart solution ?

**Comment by PaulRBerg** [comment]
I am throwing an error for what to do when the MNEMONIC is missing: `Error: Please set your MNEMONIC in a .env file`. I don't know how I could make that any clearer.

> When running the tests with hardhat, no mnemonic is needed, so should not cause any error if env is not set.

Actually a mnemonic is needed, at least the way I set up in the template. See [lines 61-63](https://github.com/paulrberg/solidity-template/blob/2de6b59757e0dd7f467d7549074329391a5d671f/hardhat.config.ts#L61-L63).

It is possible to not pass the mnemonic to the Hardhat network. But that's not good practice. Hardhat would use its default accounts, thus the tests would run against a different set of accounts across different chains (e.g. you may want to test against Ganache too). In certain cases, such as [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612), this can lead to non-deterministic test results.

> Any idea for a smart solution ?

Find another auditor? It's a bad omen that they don't have the patience to create a `.env` file. We're in crypto world here. Keeping secrets out of version control is what all of us should do.

Selectively loading the Hardhat config based on whether the MNEMONIC is set or not might:

1. Break Hardhat (see closed PR #60).
2. Complicate the template too much.
3. Lead to non-deterministic test results, as explained above.
